course5/ack/main.cc

```diff
@@ -0,0 +1,186 @@
+        ("files-without-matches,L", po::bool_switch(&fname_wo_matches), "Only print the filenames of files that do NOT match.")
+        ("no-recurse,n", po::bool_switch(&no_recurse), "No descending into subdirectories.")
+        ("cow", po::bool_switch(&cow), "Moo");
+    ;
+
+    po::positional_options_description p;
+    p.add("pattern", 1).add("search_paths", -1);
+
+    po::variables_map vm;
+    po::store(po::command_line_parser(argc, argv).options(desc).positional(p).run(), vm);
+    po::notify(vm);
+
+    if (vm.count("help")) {
+        std::cout << desc << std::endl;
+        return 1;
+    }
+
+    std::vector<std::string> input_search_paths;
+    std::vector<bfs::directory_entry> paths_to_process;
+    if (vm.count("search_paths"))
+    {
+        input_search_paths = vm["search_paths"].as<std::vector<std::string>>();
+    }
+    else
+    {
+        input_search_paths = {"."};
+    }
+    for (std::string path : input_search_paths)
+    {
+        if (!bfs::exists(path))
+        {
+            throw std::runtime_error("Path does not exist: " + path);
+        }
+        else if (bfs::is_regular_file(path))
+        {
+            paths_to_process.emplace_back(path);
+        }
+        else if (bfs::is_directory(path) && no_recurse)
+        {
+            for (bfs::directory_entry& bfs_path : bfs::directory_iterator(path))
+            {
+                if (bfs::is_regular_file(bfs_path))
+                    paths_to_process.push_back(bfs_path);
+            }
+        }
+        else if (bfs::is_directory(path))
+        {
+            for (bfs::directory_entry& bfs_path : bfs::recursive_directory_iterator(path))
+            {
+                if (bfs::is_regular_file(bfs_path))
+                    paths_to_process.push_back(bfs_path);
+            }
+        }
+        else
+        {
+            throw std::runtime_error("Path is not a regular file or a directory: " + path);
+        }
+    }
+
+    std::regex search_pattern;
+    if (vm.count("pattern"))
+    {
+        search_pattern = std::regex(vm["pattern"].as<std::string>());
+    }
+    else if (! cow)
+    {
+        std::cout << desc << std::endl;
+    }
+
+    for (bfs::directory_entry& bfs_path : paths_to_process)
+    {
+        std::vector<std::string> match_buffer;
+        find_matching_lines(bfs_path, search_pattern, match_buffer);
+        if (fname_w_matches)
+        {
+            if (match_buffer.size())
+                write(std::cout, bfs_path.path().string());
+        }
+        else if (fname_wo_matches)
+        {
+            if (!match_buffer.size())
+                write(std::cout, bfs_path.path().string());
+        }
+        else if (match_buffer.size())
+        {
+            write(std::cout, bfs_path.path().string());
+            write(std::cout, match_buffer);
+            write(std::cout, "\n");
+        }
+    }
+
+    if (cow)
+    {
+        std::vector<std::string> cow_pic = {
```

Comment:
This should be ```const char[] cow_pic = ...```, and the data would be static in static area.
When using vector, the data still sits in static area, but then, in the vector's constructor, a heap buffer is allocated and data copied.